src/main/scala/com/comcast/xfinity/sirius/uberstore/data/UberDataFileReader.scala

```diff
@@ -0,0 +1,169 @@
+package com.comcast.xfinity.sirius.uberstore.data
+
+import java.io.{BufferedInputStream, DataInputStream, EOFException, InputStream, RandomAccessFile}
+import java.nio.ByteBuffer
+import java.nio.channels.FileChannel
+
+import scala.annotation.tailrec
+
+/**
+  * Trait providing abstraction of low level read operations to an Uberstore
+  */
+trait UberDataFileReader extends AutoCloseable {
+  /**
+    * Returns the current offset within the Uberstore
+    */
+  def offset(): Long
+
+  /**
+    * Reads the next byte from the file
+    * @return the value of the next byte in the range 0 to 255, or -1 if EOF
+    */
+  def read(): Int
+
+  /**
+    * Reads the next 32-bit Integer from the file
+    * @return the 32-bit Integer
+    * @throws EOFException if EOF
+    */
+  def readInt(): Int
+
+  /**
+    * Reads the next 64-bit Long from the file
+    * @return the 64-bit Long
+    * @throws EOFException if EOF
+    */
+  def readLong(): Long
+
+  /**
+    * Populates the entire array with the next bytes from the file
+    * @param array to populate
+    * @throws EOFException if EOF
+    */
+  def readFully(array: Array[Byte]): Unit
+
+  /**
+    * Helper function that attempts to read the next 32-bit Integer from the file
+    * @return [[scala.Some]] if could successfully read the next 32-bit Integer, otherwise [[scala.None]]
+    * @throws EOFException if can only partially read the 32-bit Integer
+    */
+  def tryReadInt(): Option[Int] = {
+    val byte1 = this.read()
+    if (byte1 < 0) {
+      return None
+    }
+    val byte2 = this.read()
+    val byte3 = this.read()
+    val byte4 = this.read()
+    if (byte2 < 0 || byte3 < 0 || byte4 < 0) {
+      throw new EOFException()
+    }
+    Some((byte1 << 24) + (byte2 << 16) + (byte3 << 8) + (byte4 << 0))
+  }
```

Comment:
So this sort of a _read if not eof_ kind of thing. I don't love having to do the bitwise operations here just because we're using a read as a test. This approach distinguishes between _there are no bytes to read_ and _there were 1-3 bytes to read_. I'm not sure we care too much about that. I wonder if we should just wrap `readInt` in a `Result` and convert it to an `Option` on the way out.

The behavior is also a little confusing, at least as stated. I'd suggest either making it very explicit that this is a `readIntIfNotEof` or throwing those details in the comments. Generally something called `tryReadInt` I would not expect to explicitly throw an exception.

The alternative is to build an explicit `eof(): Boolean` function into each of the implementations for `UberDataFile` readers.

Thoughts?

Reply:
IMO I think it's worthwhile to distinguish between actually being at the end of the file vs. having a partial read as the latter would indicate a corrupt Uberstore.

I looked into having a separate `eof()` method but the issue with that was the lack of a good way of peeking into the contents of the stream without also advancing the stream.  The alternative would be to constantly check the size of the file which appeared to add a lot of overhead (I don't have that number on hand).  I can run a benchmark on setting a mark on the stream and see what that does.

Reply:
The mark/reset approach on the `InputStream` worked fairly well so I've moved to using an `eof` method.  Raw I/O ops were _slightly_ slower in a synthetic benchmark but bootstrapping a full XtvAPI didn't take any longer.